File: ts/indicators/tim-tillson/t2-exponential-moving-average/t2-exponential-moving-average.ts
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorType } from '../../core/indicator-type';
import { LineIndicator } from '../../core/line-indicator';
import { OutputType } from '../../core/outputs/output-type';
import { T2ExponentialMovingAverageOutput } from './t2-exponential-moving-average-output';
import { T2ExponentialMovingAverageLengthParams } from './t2-exponential-moving-average-params';
import { T2ExponentialMovingAverageSmoothingFactorParams } from './t2-exponential-moving-average-params';

const guardLength = (object: any): object is T2ExponentialMovingAverageLengthParams => 'length' in object;

// https://store.traders.com/-v16-c01-005smo-pdf.html

/** The __T2 Exponential Moving Average__ (__T2__, __T2EMA__) is a smoothing indicator with less lag than
 * a straight exponential moving average.
 *
 * In filter theory parlance, __T2__ is a four-pole non-linear Kalman filter.
 *
 * The __T2__ was developed by Tim Tillson and is described in the article:
 *
 * ❶ Technical Analysis of Stocks & Commodities v.16:1 (33-37), Smoothing Techniques For More Accurate Signals.
 *
 * The calculation is as follows:
 *
 * EMA¹ᵢ = EMA(Pᵢ) = αPᵢ + (1-α)EMA¹ᵢ₋₁ = EMA¹ᵢ₋₁ + α(Pᵢ - EMA¹ᵢ₋₁), 0 < α ≤ 1
 *
 * EMA²ᵢ = EMA(EMA¹ᵢ) = αEMA¹ᵢ + (1-α)EMA²ᵢ₋₁ = EMA²ᵢ₋₁ + α(EMA¹ᵢ - EMA²ᵢ₋₁), 0 < α ≤ 1
 *
 * GDᵛᵢ = (1+ν)EMA¹ᵢ - νEMA²ᵢ = EMA¹ᵢ + ν(EMA¹ᵢ - EMA²ᵢ), 0 < ν ≤ 1
 *
 * T2ᵢ = GDᵛᵢ(GDᵛᵢ)
 *
 * where _GD_ stands for 'Generalized DEMA' with 'volume' ν. The default value of __ν__ is 0.7.
 *
 * When _ν=0_, GD is just an EMA, and when _ν=1_, GD is DEMA. In between, GD is a cooler DEMA.
 *
 * If _x_ stands for the action of running a time series through an EMA,
 * _ƒ_ is our formula for Generalized Dema with 'volume' ν:
 *
 *     ƒ = (1+ν)x -νx²
 *
 * Running the filter though itself twice is equivalent to squaring _ƒ_:
 *
 *     ν²x⁴ - 2ν(1+ν)x³ + (1+ν)²x²
 *
 * The Metastock code for _T2_ is:
 *
 *    e1=Mov(P,periods,E)
 *
 *    e2=Mov(e1,periods,E)
 *
 *    e3=Mov(e2,periods,E)
 *
 *    e4=Mov(e3,periods,E)
 *
 *    c1=ν²
 *
 *    c2=-2ν(1+ν)
 *
 *    c3=(1+ν)²
 *
 *    t2=c1*e4+c2*e3+c3*e2
 *
 * The very first EMA value (the seed for subsequent values) is calculated differently.
 * This implementation allows for two algorithms for this seed.
 *
 * ❶ Use a simple average of the first 'period'. This is the most widely documented approach.
 *
 * ❷ Use first sample value as a seed. This is used in Metastock.
 */
export class T2ExponentialMovingAverage extends LineIndicator {
  private readonly smoothingFactor: number;
  private readonly firstIsAverage: boolean;
  private readonly length: number;
  private readonly length2: number;
  private readonly length3: number;
  private readonly length4: number;
  private readonly c1: number;
  private readonly c2: number;
  private readonly c3: number;
  private sum = 0;
  private count = 0;
  private ema1 = 0;
  private ema2 = 0;
  private ema3 = 0;
  private ema4 = 0;

  /**
   * Constructs an instance given a length in samples or a smoothing factor in [0, 1].
   **/
  public constructor(params: T2ExponentialMovingAverageLengthParams | T2ExponentialMovingAverageSmoothingFactorParams){
    super();
    const epsilon = 0.00000001;
    let len;
    let v;
    let mn;
    if (guardLength(params)) {
      const p = params as T2ExponentialMovingAverageLengthParams;
      len = Math.floor(p.length);
      if (len < 2) {
        throw new Error('length should be greater than 1');
      }

      v = p.volumeFactor;
      if (v < 0 || v > 1) {
        throw new Error('volume factor should be in range [0, 1]');
      }

      this.firstIsAverage = p.firstIsAverage;
      this.length = len;
      this.smoothingFactor = 2 / (len + 1);
      mn = 't2('.concat(len.toString(), ', ', v.toFixed(8),
        componentTripleMnemonic(p.barComponent, p.quoteComponent, p.tradeComponent), ')');
    } else {
      const p = params as T2ExponentialMovingAverageSmoothingFactorParams;
      if (p.smoothingFactor < 0 || p.smoothingFactor > 1) {
        throw new Error('smoothing factor should be in range [0, 1]');
      }

      v = p.volumeFactor;
      if (v < 0 || v > 1) {
        throw new Error('volume factor should be in range [0, 1]');
      }

      let sf = p.smoothingFactor;
      if (sf < epsilon) {
        sf = epsilon;
      }

      this.smoothingFactor = sf;
      this.length = Math.round(2 / sf) - 1;
      this.firstIsAverage = p.firstIsAverage;
      mn = 't2('.concat(this.length.toString(), ', ', sf.toFixed(8), ', ', v.toFixed(8),
        componentTripleMnemonic(p.barComponent, p.quoteComponent, p.tradeComponent), ')');
    }

    const v1 = v + 1;
    this.c1 = v * v;
    this.c2 = -2 * v * v1;
    this.c3 = v1 * v1;

    const l = this.length;
    this.length2 = l * 2 - 1;
    this.length3 = l * 3 - 2;
    this.length4 = l * 4 - 3;

    this.mnemonic = mn;
    this.description = 'T2 exponential moving average ' + mn;
    this.barComponent = params.barComponent;
    this.quoteComponent = params.quoteComponent;
    this.tradeComponent = params.tradeComponent;
    this.primed = false;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return {
      type: IndicatorType.T2ExponentialMovingAverage,
      mnemonic: this.mnemonic,
      description: this.description,
      outputs: [{
        kind: T2ExponentialMovingAverageOutput.T2ExponentialMovingAverageValue,
        type: OutputType.Scalar,
        mnemonic: this.mnemonic,
        description: this.description,
      }],
    };
  }

  /** Updates the value of the indicator given the next sample. */
  public update(sample: number): number {
    if (Number.isNaN(sample)) {
      return sample;
    }

    const sf = this.smoothingFactor;
    if (this.primed) {
      this.ema1 += (sample - this.ema1) * sf;
      this.ema2 += (this.ema1 - this.ema2) * sf;
      this.ema3 += (this.ema2 - this.ema3) * sf;
      this.ema4 += (this.ema3 - this.ema4) * sf;
      return this.c1 * this.ema4 + this.c2 * this.ema3 + this.c3 * this.ema2;
    }

    // Not primed.
    ++this.count;
    if (this.firstIsAverage) { // First is the simple average.
      if (this.count === 1) {
        this.sum = sample;
      } else if (this.length >= this.count) {
        this.sum += sample;
        if (this.length === this.count) {
          this.ema1 = this.sum / this.length;
          this.sum = this.ema1;
        }
      } else if (this.length2 >= this.count) {
        this.ema1 += (sample - this.ema1) * sf;
        this.sum += this.ema1;
        if (this.length2 === this.count) {
          this.ema2 = this.sum / this.length;
          this.sum = this.ema2;
        }
      } else if (this.length3 >= this.count) {
        this.ema1 += (sample - this.ema1) * sf;
        this.ema2 += (this.ema1 - this.ema2) * sf;
        this.sum += this.ema2;
        if (this.length3 === this.count) {
          this.ema3 = this.sum / this.length;
          this.sum = this.ema3;
        }
      } else { // if (this.length4 >= this.count) {
        this.ema1 += (sample - this.ema1) * sf;
        this.ema2 += (this.ema1 - this.ema2) * sf;
        this.ema3 += (this.ema2 - this.ema3) * sf;
        this.sum += this.ema3;
        if (this.length4 === this.count) {
          this.primed = true;
          this.ema4 = this.sum / this.length;
          return this.c1 * this.ema4 + this.c2 * this.ema3 + this.c3 * this.ema2;
        }
      }
    } else { // firstIsAverage is false, Metastock case.
      if (this.count === 1) {
        this.ema1 = sample;
      } else if (this.length >= this.count) {
        this.ema1 += (sample - this.ema1) * sf;
        if (this.length === this.count) {
          this.ema2 = this.ema1;
        }
      } else if (this.length2 >= this.count) {
        this.ema1 += (sample - this.ema1) * sf;
        this.ema2 += (this.ema1 - this.ema2) * sf;
        if (this.length2 === this.count) {
          this.ema3 = this.ema2;
        }
      } else if (this.length3 >= this.count) {
        this.ema1 += (sample - this.ema1) * sf;
        this.ema2 += (this.ema1 - this.ema2) * sf;
        this.ema3 += (this.ema2 - this.ema3) * sf;
        if (this.length3 === this.count) {
          this.ema4 = this.ema3;
        }
      } else { // if (this.length4 >= this.count) {
        this.ema1 += (sample - this.ema1) * sf;
        this.ema2 += (this.ema1 - this.ema2) * sf;
        this.ema3 += (this.ema2 - this.ema3) * sf;
        this.ema4 += (this.ema3 - this.ema4) * sf;
        if (this.length4 === this.count) {
          this.primed = true;
          return this.c1 * this.ema4 + this.c2 * this.ema3 + this.c3 * this.ema2;
        }
      }
    }

    return Number.NaN;
  }
}

File: go/indicators/timtillson/t2exponentialmovingaverage/t2exponentialmovingaverage.go
package t2exponentialmovingaverage

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// https://store.traders.com/-v16-c01-005smo-pdf.html

// T2ExponentialMovingAverage (T2 Exponential Moving Average, T2, T2EMA)
// is a smoothing indicator with less lag than a straight exponential moving average.
//
// In filter theory parlance, T2 is a four-pole non-linear Kalman filter.
//
// The T2 was developed by Tim Tillson and is described in the article:
//
//	❶ Technical Analysis of Stocks & Commodities v.16:1 (33-37), Smoothing Techniques For More Accurate Signals.
//
// The calculation is as follows:
//
//	EMA¹ᵢ = EMA(Pᵢ) = αPᵢ + (1-α)EMA¹ᵢ₋₁ = EMA¹ᵢ₋₁ + α(Pᵢ - EMA¹ᵢ₋₁), 0 < α ≤ 1
//	EMA²ᵢ = EMA(EMA¹ᵢ) = αEMA¹ᵢ + (1-α)EMA²ᵢ₋₁ = EMA²ᵢ₋₁ + α(EMA¹ᵢ - EMA²ᵢ₋₁), 0 < α ≤ 1
//	GDᵛᵢ = (1+ν)EMA¹ᵢ - νEMA²ᵢ = EMA¹ᵢ + ν(EMA¹ᵢ - EMA²ᵢ), 0 < ν ≤ 1
//	T2ᵢ = GDᵛᵢ(GDᵛᵢ)
//
// where GD stands for 'Generalized DEMA' with 'volume' ν. The default value of ν is 0.7.
// When ν=0, GD is just an EMA, and when ν=1, GD is DEMA. In between, GD is a cooler DEMA.
//
// If x stands for the action of running a time series through an EMA,
// ƒ is our formula for Generalized Dema with 'volume' ν:
//
//	ƒ = (1+ν)x -νx²
//
// Running the filter though itself twice is equivalent to squaring ƒ:
//
//	ν²x⁴ - 2ν(1+ν)x³ + (1+ν)²x²
//
// The Metastock code for T2 is:
//
//	e1=Mov(P,periods,E)
//	e2=Mov(e1,periods,E)
//	e3=Mov(e2,periods,E)
//	e4=Mov(e3,periods,E)
//	c1=ν²
//	c2=-2ν(1+ν)
//	c3=(1+ν)²
//	t2=c1*e4+c2*e3+c3*e2
//
// The very first EMA value (the seed for subsequent values) is calculated differently.
// This implementation allows for two algorithms for this seed.
//
//	❶ Use a simple average of the first 'period'. This is the most widely documented approach.
//	❷ Use first sample value as a seed. This is used in Metastock.
type T2ExponentialMovingAverage struct {
	mu sync.RWMutex
	core.LineIndicator
	smoothingFactor float64
	c1              float64
	c2              float64
	c3              float64
	sum             float64
	ema1            float64
	ema2            float64
	ema3            float64
	ema4            float64
	length          int
	length2         int
	length3         int
	length4         int
	count           int
	firstIsAverage  bool
	primed          bool
}

// NewT2ExponentialMovingAverageLength returns an instnce of the indicator
// created using supplied parameters based on length.
func NewT2ExponentialMovingAverageLength(
	p *T2ExponentialMovingAverageLengthParams,
) (*T2ExponentialMovingAverage, error) {
	return newT2ExponentialMovingAverage(p.Length, math.NaN(), p.VolumeFactor,
		p.FirstIsAverage, p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

// NewT2ExponentialMovingAverageSmoothingFactor returns an instnce of the indicator
// created using supplied parameters based on smoothing factor.
func NewT2ExponentialMovingAverageSmoothingFactor(
	p *T2ExponentialMovingAverageSmoothingFactorParams,
) (*T2ExponentialMovingAverage, error) {
	return newT2ExponentialMovingAverage(0, p.SmoothingFactor, p.VolumeFactor,
		p.FirstIsAverage, p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

//nolint:funlen,cyclop
func newT2ExponentialMovingAverage(length int, alpha float64, v float64, firstIsAverage bool,
	bc entities.BarComponent, qc entities.QuoteComponent, tc entities.TradeComponent,
) (*T2ExponentialMovingAverage, error) {
	const (
		invalid = "invalid t2 exponential moving average parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
		fmtnl   = "t2(%d, %.8f%s)"
		fmtna   = "t2(%d, %.8f, %.8f%s)"
		minlen  = 2
		two     = 2.
		three   = 3
		four    = 4
		epsilon = 0.00000001
	)

	var (
		mnemonic  string
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if v < 0. || v > 1. {
		return nil, fmt.Errorf(fmts, invalid, "volume factor should be in range [0, 1]")
	}

	// Resolve defaults for component functions.
	// A zero value means "use default, don't show in mnemonic".
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	if math.IsNaN(alpha) {
		if length < minlen {
			return nil, fmt.Errorf(fmts, invalid, "length should be greater than 1")
		}

		alpha = two / float64(1+length)
		mnemonic = fmt.Sprintf(fmtnl, length, v, core.ComponentTripleMnemonic(bc, qc, tc))
	} else {
		if alpha < 0. || alpha > 1. {
			return nil, fmt.Errorf(fmts, invalid, "smoothing factor should be in range [0, 1]")
		}

		if alpha < epsilon {
			alpha = epsilon
		}

		length = int(math.Round(two/alpha)) - 1
		mnemonic = fmt.Sprintf(fmtna, length, alpha, v, core.ComponentTripleMnemonic(bc, qc, tc))
	}

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	desc := "T2 exponential moving average " + mnemonic

	v1 := v + 1
	c1 := v * v
	c2 := -two * v * v1
	c3 := v1 * v1

	t2 := &T2ExponentialMovingAverage{
		smoothingFactor: alpha,
		c1:              c1,
		c2:              c2,
		c3:              c3,
		length:          length,
		length2:         int(two)*length - 1,
		length3:         three*length - int(two),
		length4:         four*length - three,
		firstIsAverage:  firstIsAverage,
	}

	t2.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, t2.Update)

	return t2, nil
}

// IsPrimed indicates whether an indicator is primed.
func (s *T2ExponentialMovingAverage) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes an output data of the indicator.
// It always has a single scalar output -- the calculated value of the T2 exponential moving average.
func (s *T2ExponentialMovingAverage) Metadata() core.Metadata {
	return core.Metadata{
		Type:        core.T2ExponentialMovingAverage,
		Mnemonic:    s.LineIndicator.Mnemonic,
		Description: s.LineIndicator.Description,
		Outputs: []outputs.Metadata{
			{
				Kind:        int(T2ExponentialMovingAverageValue),
				Type:        outputs.ScalarType,
				Mnemonic:    s.LineIndicator.Mnemonic,
				Description: s.LineIndicator.Description,
			},
		},
	}
}

// Update updates the value of the indicator given the next sample.
func (s *T2ExponentialMovingAverage) Update(sample float64) float64 { //nolint:cyclop,funlen,gocognit
	if math.IsNaN(sample) {
		return sample
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	sf := s.smoothingFactor

	if s.primed {
		v1 := s.ema1
		v2 := s.ema2
		v3 := s.ema3
		v4 := s.ema4
		v1 += (sample - v1) * sf
		v2 += (v1 - v2) * sf
		v3 += (v2 - v3) * sf
		v4 += (v3 - v4) * sf
		s.ema1 = v1
		s.ema2 = v2
		s.ema3 = v3
		s.ema4 = v4

		return s.c1*v4 + s.c2*v3 + s.c3*v2
	}

	s.count++
	if s.firstIsAverage { //nolint:nestif
		if s.count == 1 {
			s.sum = sample
		} else if s.length >= s.count {
			s.sum += sample
			if s.length == s.count {
				s.ema1 = s.sum / float64(s.length)
				s.sum = s.ema1
			}
		} else if s.length2 >= s.count {
			s.ema1 += (sample - s.ema1) * sf
			s.sum += s.ema1

			if s.length2 == s.count {
				s.ema2 = s.sum / float64(s.length)
				s.sum = s.ema2
			}
		} else if s.length3 >= s.count {
			s.ema1 += (sample - s.ema1) * sf
			s.ema2 += (s.ema1 - s.ema2) * sf
			s.sum += s.ema2

			if s.length3 == s.count {
				s.ema3 = s.sum / float64(s.length)
				s.sum = s.ema3
			}
		} else { // if s.length4 >= s.count {
			s.ema1 += (sample - s.ema1) * sf
			s.ema2 += (s.ema1 - s.ema2) * sf
			s.ema3 += (s.ema2 - s.ema3) * sf
			s.sum += s.ema3

			if s.length4 == s.count {
				s.primed = true
				s.ema4 = s.sum / float64(s.length)

				return s.c1*s.ema4 + s.c2*s.ema3 + s.c3*s.ema2
			}
		}
	} else { // Metastock
		if s.count == 1 {
			s.ema1 = sample
		} else if s.length >= s.count {
			s.ema1 += (sample - s.ema1) * sf
			if s.length == s.count {
				s.ema2 = s.ema1
			}
		} else if s.length2 >= s.count {
			s.ema1 += (sample - s.ema1) * sf
			s.ema2 += (s.ema1 - s.ema2) * sf

			if s.length2 == s.count {
				s.ema3 = s.ema2
			}
		} else if s.length3 >= s.count {
			s.ema1 += (sample - s.ema1) * sf
			s.ema2 += (s.ema1 - s.ema2) * sf
			s.ema3 += (s.ema2 - s.ema3) * sf

			if s.length3 == s.count {
				s.ema4 = s.ema3
			}
		} else { // if s.length4 >= s.count {
			s.ema1 += (sample - s.ema1) * sf
			s.ema2 += (s.ema1 - s.ema2) * sf
			s.ema3 += (s.ema2 - s.ema3) * sf
			s.ema4 += (s.ema3 - s.ema4) * sf

			if s.length4 == s.count {
				s.primed = true

				return s.c1*s.ema4 + s.c2*s.ema3 + s.c3*s.ema2
			}
		}
	}

	return math.NaN()
}
